### PR TITLE
Adds plane calibration

### PR DESCRIPTION
**Pull request by narora1**

### CODE REVIEW

File: robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h [hunk 1]
```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (C) 2015 Fetch Robotics Inc.
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h]: 2015-2017

File: robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+namespace robot_calibration
+{
+
+/**
+ *  \brief Model of a camera on another camera chain.
+ */
+struct CameraToCameraError
+{
+    /**
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h]: this whole thing should only be indented by 2 -- appears to be 4

File: robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+     *  \brief Error block for calibrating a 3d camera to a single arm.
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h]: update description -- no arm involved

File: robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      // Project the first camera observations
+      std::vector <geometry_msgs::PointStamped> camera1_pts =
+          camera1_model_->project(data_, *offsets_);
+
+      // Project the second camera estimation
+      std::vector <geometry_msgs::PointStamped> camera2_pts =
+          camera2_model_->project(data_, *offsets_);
+
+      // Calculate plane normals using SVD
+
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h]: remove extra newline

File: robot_calibration/src/optimizer.cpp [hunk 5]
```diff
@@ -238,6 +238,59 @@ int Optimizer::optimize(OptimizationParams& params,
...
                                   free_params);
       }
+      if (params.error_blocks[j].type == "camera_to_camera")
+      {
+        // This error block can process data generated by the PlaneFinder,
+        // CheckerboardFinder, or any other finder that can sample the pose
+        // of one or more data points in two camera frames.
+        // This also finds the a plane that fits the points and minimizes
+        // over the normals
+
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: remove extra newline

File: robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+          index = k;
+          break;
+        }
+      }
+
+      if (index == -1)
+      {
+        std::cerr << "Sensor name doesn't match any of the existing finders" << std::endl;
+        return 0;
+      }
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h]: L119-133 doesn't seem like it should be in this function -- it should be around L272 in optimizer.cpp right after we extract camera1_model -- if we haven't found camera2_model -- we should abort creating the block. returning 0 here give you a null pointer -- which is probably going to blow shit up.

Comment (narora1) [reply robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h]: I updated the one in optimizer.cpp but the other error blocks also have the check. Should I remove it from them as well??

**Comment by mikeferguson** [comment]
Updates look good -- but you added a calibration_poses.bag file in the scripts folder -- please remove that in another commit -- I can squash everything when I merge

File: robot_calibration/src/plane_finder.cpp [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+  nh.param<std::string>("topic", topic_name, "/points");
+  subscriber_ = nh.subscribe(topic_name,
+                             1,
+                             &PlaneFinder::cameraCallback,
+                             this);
+
+  nh.param<std::string>("camera_sensor_name", camera_sensor_name_, "camera");
+  nh.param<double>("points_max", points_max_, 60);
+  nh.param<double>("min_y", min_x_, -2.0);
+  nh.param<double>("max_y", max_x_, 2.0);
```
Comment (mikeferguson) [inline robot_calibration/src/plane_finder.cpp]: the should be min_x/max_x for param name, no?

Comment (mikeferguson) [reply robot_calibration/src/plane_finder.cpp]: also, looks like min/max_x_ aren't defined anywhere?

File: robot_calibration/src/plane_finder.cpp [hunk 8]
```diff
@@ -143,13 +144,14 @@ bool PlaneFinder::find(robot_calibration_msgs::CalibrationData * msg)
...
 
-    // Remove the points immediately in front of te camera in the point cloud
+    // Remove the points immediately in front of the camera in the point cloud
     // NOTE : This is to handle sensors that publish zeros instead of NaNs in the point cloud
     if (p.point.z == 0)
       continue;
 
     // Transformed point in the transformed frame (default is base_link))
-    if (p_out.point.z < min_z_ or p_out.point.z > max_z_ or p_out.point.y < min_y or p_out.point.y > max_y)
+    if (p_out.point.y < min_y or p_out.point.y > max_y or p_out.point.y < min_y or p_out.point.y > max_y or
```
Comment (mikeferguson) [inline robot_calibration/src/plane_finder.cpp]: min_y_ -- you're missing underscore

Comment (mikeferguson) [reply robot_calibration/src/plane_finder.cpp]: wait -- "or" ? That's not even C++?

File: robot_calibration/include/robot_calibration/capture/plane_finder.h [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  sensor_msgs::PointCloud2 cloud_;
+  DepthCameraInfoManager depth_camera_manager_;
+
+  std::string camera_sensor_name_;
+  double points_max_;
+  double min_z_;
+  double max_z_;
+  double min_y_;
+  double max_y_;
+  double transform_frame_;
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/capture/plane_finder.h]: this is supposed to be a string -- doesn't compile as a double!

File: robot_calibration/src/plane_finder.cpp (hunk 8 above)
Comment (gavanderhoorn) [reply robot_calibration/src/plane_finder.cpp]: > wait -- "or" ? That's not even C++?

It actually is :) See [cppreference/Logical operators](http://en.cppreference.com/w/cpp/language/operator_logical).

Comment (mikeferguson) [reply robot_calibration/src/plane_finder.cpp]: @gavanderhoorn thanks. Wow. In something like 15 years of C/C++ work, I don't think I've ever used that, or even knew it existed!

That said -- style guide says don't do it: https://google.github.io/styleguide/cppguide.html#Boolean_Expressions

And one other note @narora1  -- this has two comparisons for Y -- I think one of those was meant to be for X.

**Pull request merged**